### PR TITLE
feat: human focus transfer domain

### PR DESCRIPTION
**Pull request by canonbrother**
###### Requirement
```
transferdomain <from> <to> <amount> <domain=auto> <domain-specific-opts...> (eg: change-addr=from-for-utxo)
```

Simplifiy transfer domain command line form
###### from

transferdomain [{"src":{"address":src,"amount":"1@DFI","domain":3},"dst":{"address":dst,"amount":"1@DFI","domain": 2}}]

###### to

~transferdomain src dst tokenAmount(eg: 1@DFI) domain(0 - auto, 1 - utxo, 2 - dvm, 3 - evm) nonce(optional)~
transferdomain src dst tokenAmount domain(eg: 1@DFI) nonce(optional)

## RPCs

- transferdomain

### CODE REVIEW

**Comment by Jouzo** [comment]
> transferdomain src dst tokenAmount(eg: 1@DFI) nonce(optional)
Are we inferring the domains ?

**Comment by canonbrother** [comment]
> > transferdomain src dst tokenAmount(eg: 1@DFI) nonce(optional)
> 
> Are we inferring the domains ?

was removed.. as just found out the domain can be justified via `from` and `to` automatically.. 
eg: `if from == bech32 && to == erc55 then domain = evm`

**Comment by Jouzo** [comment]
Shouldn't we add an additional optional domain to be future proof, i.e. for when we add support to UTXO ?

**Comment by canonbrother** [comment]
> Shouldn't we add an additional optional domain to be future proof, i.e. for when we add support to UTXO ?

hmm.. yea.. finishing up the basic first.. now impl the utox stuff 😬

File: src/dfi/consensus/xvm.h [hunk 1]
```diff
@@ -11,8 +11,7 @@ struct CEvmTxMessage;
 struct CTransferDomainMessage;
 
 enum class VMDomain : uint8_t {
-    NONE = 0x00,
-    // UTXO Reserved
+    AUTO = 0x00,
```
Comment (prasannavl) [inline src/dfi/consensus/xvm.h:14]: I'd still keep this as `None` to represent the Unknown state on the enum. 

The auto is just purely on the RPC side, where it shifts unknown state into a deterministic known state. Wouldn't add an auto state to the protocol message itself.

File: src/dfi/rpc_accounts.cpp [hunk 2]
```diff
@@ -2117,77 +2117,96 @@ UniValue sendtokenstoaddress(const JSONRPCRequest &request) {
 
 UniValue transferdomain(const JSONRPCRequest &request) {
     auto pwallet = GetWallet(request);
-    // TODO: Add support for non-JSON parameteric input that's human friendly and intuitive
-    RPCHelpMan{
+
+    if (!request.params[0].isArray()) {
+        RPCHelpMan{
```
Comment (prasannavl) [inline src/dfi/rpc_accounts.cpp:2122]: Help params need to be outside of the if branch and with both help messages in the same unit. 
Only the final `Check` call inside.

Comment (canonbrother) [reply src/dfi/rpc_accounts.cpp]: not sure i'm doing it correctly
refined https://github.com/DeFiCh/ain/pull/2679/commits/9e23c237d15170b0f7353851384b9600a1705456

Comment (canonbrother) [reply src/dfi/rpc_accounts.cpp]: custom made RPCHelpMan https://github.com/DeFiCh/ain/pull/2679/commits/480b0806b7c27105a32ee0f0418c04efcabf8728
```
transferdomain
Creates (and submits to local node and network) a tx to transfer assets across domains. DVM to EVM/EVM to DVM, etc.

Arguments:
1. array | string                       (json array | string, required) A json array of src and dst json objects
    [
    {                        (json object)
        "src": {               (json object) Source arguments
        "address": "str",    (string, required) Source address
        "amount": "str",     (string, required) Amount transfered, the value is amount in amount@token format
        "domain": n,         (numeric, required) Domain of source: 2 - DVM, 3 - EVM
        },
        "dst": {               (json object) Destination arguments
        "address": "str",    (string, required) Destination address
        "amount": "str",     (string, required) Amount transfered, the value is amount in amount@token format
        "domain": n,         (numeric, required) Domain of source: 2 - DVM, 3 - EVM
        },
        "nonce": n,            (numeric) Optional parameter to specify the transaction nonce
    },
    ...
    ]
   from                         (string, required) the source address of sender
2. to                           (string, required) the destination address of sender
3. tokenAmount                  (string, required) in amount@token format
4. nonce

Result:
"hash"                  (string) The hex-encoded hash of broadcasted transaction

Examples:
> defi-cli transferdomain '[{"src":{"address":"<DFI_address>", "amount":"1.0@DFI", "domain": 2}, "dst":{"address":"<ETH_address>", "amount":"1.0@DFI", "domain": 3}}]'
> defi-cli transferdomain '[{"src":{"address":"<ETH_address>", "amount":"1.0@DFI", "domain": 3}, "dst":{"address":"<DFI_address>", "amount":"1.0@DFI", "domain": 2}}]'
> defi-cli transferdomain "from" "to" "100@DFI"
> defi-cli transferdomain "from", "to", 100@BTC 2"
```

**Comment by canonbrother** [comment]
reverted utxo updates
checkpoint at https://github.com/DeFiCh/ain/pull/2679/commits/9e0c55773fe160ef31a1bbd782a63a790b1b3e0e
prefix at src/dst if needed since utxo is not supported

**Comment by canonbrother** [comment]
⚠️ On hold PR.. `transferdomain` not working due to RPCHelpMan arg checks.. will check again

**Pull request closed without merging**